Wilfred Morgan: Javascript IV Project

wmemorgan opened the pull request:
# Javascript IV Project Submission

## Project Objectives
### MVP
* [x] `prototype-refactor` - Take existing code and make it modern.
* [x] `lambda-classes` - We need a roster of Lambda School personnel. Build it!

### Stretch Problem
* [x] Extend the functionality of the Student by adding a prop called grade and setting it equal to a number between 1-100.
* [x] Now that our students have a grade build out a method on the Instructor (this will be used by _BOTH_ instructors and PM's) that will randomly add or subtract points to a student's grade. _Math.random_ will help.
* [x] Add a graduate method to a student.
  * [x] This method, when called, will check the grade of the student and see if they're ready to graduate from Lambda School
  * [x] If the student's grade is above a 70% let them graduate! Otherwise go back to grading their assignments to increase their score.

---
## Summary

### Blocks
Ran into a stumbling block with listing a student's favorite subjects. This required enumerating the `Student.favSubjects` array which I initially did using console.log(). This caused me to get `undefined` during testing.
 
### Breakthroughs
After reviewing the issue with @JoeBugajski, he pointed out that creating functions that just contain `console.log()` commands but do not return an actual value will not work if you try to invoke the function using `console.log()`.

I realized that for functions I should always make sure to actually `return` value (instead of just console logging function results).

---
## Additional Notes

wmemorgan committed:
Initialize project

wmemorgan committed:
Refactor all video game objects into classes

wmemorgan committed:
Refactor game sequence functions

wmemorgan committed:
Create classes for Lambda School personnel

wmemorgan committed:
Fix typo in "lambda-classes" section

wmemorgan committed:
Create school personnel

wmemorgan committed:
Fix missing gender properties

wmemorgan committed:
Display school personnel

wmemorgan committed:
Add missing method parameter for PM grade method

Update project manager stand up message

JoeBugajski commented on assignments/lambda-classes.js:
```diff
@@ -1 +1,222 @@
+    this.gender = attributes.gender
+  }
+  speak() {
+    console.log(`Hello my name is ${this.name}, I am from ${this.location}.`)
```
These methods should probably be returning a string, rather than console logging. That's why you were getting all those undefined messages, because you were trying to console log a console log. 

```suggestion
    return `Hello my name is ${this.name}, I am from ${this.location}.`
```

wmemorgan replied:
Yes, I was wondering about that exact same thing. The instructions said to _log out_ the messages but I was thinking that returning a string should be the proper method. 

I'll update the methods and have it return the strings.

Thanks Joe!

wmemorgan replied:
All suggested changes have been applied. Let me know if any changes are needed. In the meantime I resolve this note. Thanks.

JoeBugajski reviewed (changes requested):
The only problem I can see is many of your methods are console logging a string, rather than returning a string. For some reason, it doesn't like console logging a console log. It's a controversial problem because the instructions steer you in the wrong direction.

wmemorgan committed:
Add return string commands for all class methods

wmemorgan committed:
Comment out console.logs from all class methods

wmemorgan committed:
Add 2 different approaches for listing a student's favorite subjects

wmemorgan committed:
Add `grade` property to Student class

Add grade assignment method to Instructor class

Add graduation verification method

Add function to test the stretch goals

wmemorgan committed:
Update gitignore file

wmemorgan committed:
Add a special Lambda School "no student left behind" conditional loop
